### PR TITLE
added json output support for stats handler

### PR DESCRIPTION
**Pull request by ramaraochavali**
PR for [issue](https://github.com/envoyproxy/envoy/issues/1862)

### CODE REVIEW

File: source/server/http/admin.cc [hunk 1]
```diff
@@ -279,11 +289,42 @@ Http::Code AdminImpl::handlerStats(const std::string&, Buffer::Instance& respons
...
 
-  for (auto stat : all_stats) {
-    response.add(fmt::format("{}: {}\n", stat.first, stat.second));
+  if (params.size() == 0) {
+    // No Arguments so use the standard
+    for (auto stat : all_stats) {
+      response.add(fmt::format("{}: {}\n", stat.first, stat.second));
+    }
+  } else {
+    std::string format = params.begin()->first;
```
Comment (ramaraochavali) [inline source/server/http/admin.cc]: Tried to use Field, FieldPtr and create json but not able to include them here.

**Comment by mattklein123** [comment]
@ramaraochavali before I dive into full review, a few things:
1) Will need DCO (see contributing guide).
2) Please get the PR compiling/passing tests.
3) Will need docs here: https://envoyproxy.github.io/envoy/operations/admin.html
4) Will need a test here https://github.com/envoyproxy/envoy/blob/master/test/integration/integration_admin_test.cc that the body returned is actually valid JSON.

**Comment by ramaraochavali** [comment]
@mattklein123  I have updated the doc and added an integration test that validates the json. I have signed the DCO in my last commit. But still it shows as failed? Do I have to commit with a new branch?
All tests are passing. Can you please review now?

**Comment by mattklein123** [comment]
@ramaraochavali every commit has to have DCO. Please just squash/rebase and force push the entire PR on current master (to fix OSX tests) and then we can review. Thank you.

**Comment by ramaraochavali** [comment]
@mattklein123 done..did a force commit as single one

File: docs/operations/admin.rst [hunk 2]
```diff
@@ -137,3 +137,7 @@ The fields are:
   output as Envoy currently has no built in histogram support and relies on statsd for
   aggregation. This command is very useful for local debugging. See :ref:`here <operations_stats>`
   for more information.
+
+  .. http:get:: /stats?format=json
+
+    Same as /stats but outputs them in json format. This can be used for programmatic access of stats.
```
Comment (mattklein123) [inline docs/operations/admin.rst]: nit: "Same as /stats the output is in JSON format"

File: source/server/http/admin.cc [hunk 3]
```diff
@@ -31,8 +31,17 @@
 #include "common/upstream/host_utility.h"
 
 #include "fmt/format.h"
+#include "rapidjson/document.h"
```
Comment (mattklein123) [inline source/server/http/admin.cc]: I would prefer that we don't like rapidjson out into the rest of the code, but it's OK for now. Can you please add above the includes:

"TODO(mattklein123): Switch to JSON interface methods and remove rapidjson dependency"

File: source/server/http/admin.cc (hunk 1 above)
Comment (mattklein123) [reply source/server/http/admin.cc]: Yeah, thanks for trying. See TODO request above.

File: source/server/http/admin.cc [hunk 4]
```diff
@@ -279,11 +290,42 @@ Http::Code AdminImpl::handlerStats(const std::string&, Buffer::Instance& respons
     all_stats.emplace(gauge->name(), gauge->value());
   }
 
-  for (auto stat : all_stats) {
-    response.add(fmt::format("{}: {}\n", stat.first, stat.second));
+  if (params.size() == 0) {
+    // No Arguments so use the standard
```
Comment (mattklein123) [inline source/server/http/admin.cc]: nit: "No arguments so use the standard output." (period end of sentence)

File: source/server/http/admin.cc [hunk 5]
```diff
@@ -279,11 +290,42 @@ Http::Code AdminImpl::handlerStats(const std::string&, Buffer::Instance& respons
...
+    for (auto stat : all_stats) {
+      response.add(fmt::format("{}: {}\n", stat.first, stat.second));
+    }
+  } else {
+    std::string format = params.begin()->first;
+    std::string formatvalue = params.begin()->second;
```
Comment (mattklein123) [inline source/server/http/admin.cc]: Please use snake case for all vars (e.g., "format_value", "stats_array" below, etc.). Won't comment on each one.

File: source/server/http/admin.cc [hunk 6]
```diff
@@ -279,11 +290,42 @@ Http::Code AdminImpl::handlerStats(const std::string&, Buffer::Instance& respons
...
+    if (format == "format" && formatvalue == "json") {
+      rapidjson::Document document;
```
Comment (mattklein123) [inline source/server/http/admin.cc]: nit: Can we move the actual JSON creation into a helper function. We will likely have other output types in the future.

File: source/server/http/admin.cc (hunk 5 above)
Comment (ramaraochavali) [reply source/server/http/admin.cc]: will change.. this habit from my lot of java coding :-)

File: source/server/http/admin.cc (hunk 6 above)
Comment (ramaraochavali) [reply source/server/http/admin.cc]: will do

**Comment by ramaraochavali** [comment]
@mattklein123 Thanks for your quick review. Addressed all the comments. Should be good to go. Let me know if further changes are required

File: docs/operations/admin.rst [hunk 7]
```diff
@@ -137,3 +137,7 @@ The fields are:
   output as Envoy currently has no built in histogram support and relies on statsd for
   aggregation. This command is very useful for local debugging. See :ref:`here <operations_stats>`
   for more information.
+
+  .. http:get:: /stats?format=json
+
+  Same as /stats the output is in JSON format. This can be used for programmatic access of stats.
```
Comment (mattklein123) [inline docs/operations/admin.rst]: Please switch first sentence to "Outputs /stats in JSON format."

File: source/server/http/admin.cc [hunk 8]
```diff
@@ -267,9 +278,11 @@ Http::Code AdminImpl::handlerServerInfo(const std::string&, Buffer::Instance& re
   return Http::Code::OK;
 }
 
-Http::Code AdminImpl::handlerStats(const std::string&, Buffer::Instance& response) {
+Http::Code AdminImpl::handlerStats(const std::string& url, Buffer::Instance& response) {
   // We currently don't support timers locally (only via statsd) so just group all the counters
   // and gauges together, alpha sort them, and spit them out.
+  Http::Code rc = Http::Code::OK;
+  Http::Utility::QueryParams params = Http::Utility::parseQueryString(url);
```
Comment (mattklein123) [inline source/server/http/admin.cc]: nit: const

File: source/server/http/admin.cc [hunk 9]
```diff
@@ -279,11 +292,47 @@ Http::Code AdminImpl::handlerStats(const std::string&, Buffer::Instance& respons
...
 
-  for (auto stat : all_stats) {
-    response.add(fmt::format("{}: {}\n", stat.first, stat.second));
+  if (params.size() == 0) {
+    // No Arguments so use the standard.
+    for (auto stat : all_stats) {
+      response.add(fmt::format("{}: {}\n", stat.first, stat.second));
+    }
+  } else {
+    std::string format_key = params.begin()->first;
```
Comment (mattklein123) [inline source/server/http/admin.cc]: nit: const for this var and the the following

File: source/server/http/admin.cc [hunk 10]
```diff
@@ -279,11 +292,47 @@ Http::Code AdminImpl::handlerStats(const std::string&, Buffer::Instance& respons
...
+    std::string format_value = params.begin()->second;
+    if (format_key == "format" && format_value == "json") {
+      std::string json_stats = statsAsJson(all_stats);
```
Comment (mattklein123) [inline source/server/http/admin.cc]: nit: don't need local var, just fold function call into the following line.

File: source/server/http/admin.h [hunk 11]
```diff
@@ -115,7 +115,7 @@ class AdminImpl : public Admin,
   void addOutlierInfo(const std::string& cluster_name,
                       const Upstream::Outlier::Detector* outlier_detector,
                       Buffer::Instance& response);
-
+  std::string statsAsJson(std::map<std::string, uint64_t> all_stats);
```
Comment (mattklein123) [inline source/server/http/admin.h]: nit: newline after this line.

**Comment by ramaraochavali** [comment]
@mattklein123 I think have addressed all comments. Have a final look once. Sorry because of lack of my recent c++ coding exp, took more time for you to review I guess. :-)

File: source/server/http/admin.h [hunk 12]
```diff
@@ -115,6 +115,7 @@ class AdminImpl : public Admin,
   void addOutlierInfo(const std::string& cluster_name,
                       const Upstream::Outlier::Detector* outlier_detector,
                       Buffer::Instance& response);
+  std::string statsAsJson(std::map<std::string, uint64_t> all_stats);
```
Comment (mattklein123) [inline source/server/http/admin.h]: oops sorry one more thing, can you make this `const std::map<std::string, uint64_t>&`. Sorry! :)

Comment (ramaraochavali) [reply source/server/http/admin.h]: No issues. committed..please check

**Pull request merged**